### PR TITLE
Feature / On Press Callback

### PR DESCRIPTION
**Pull request by malkovro**
## Changes

- Add onPress registering on addToWallet button pressed

Example app adjusted to print to the console if button is pressed

### CODE REVIEW

File: ios/Classes/SwiftAddToWalletPlugin.swift [hunk 1]
```diff
@@ -34,17 +37,22 @@ class PKAddPassButtonNativeView: NSObject, FlutterPlatformView {
...
-        binaryMessenger messenger: FlutterBinaryMessenger?
+        binaryMessenger messenger: FlutterBinaryMessenger?,
+        channel: FlutterMethodChannel
     ) {
         _view = UIView()
         _pass = args["pass"] as! FlutterStandardTypedData
         _width = args["width"] as? CGFloat ?? 140
         _height = args["height"] as? CGFloat ?? 30
+        _key = args["key"] as! String
+        _channel = channel
```
Comment (malkovro) [inline ios/Classes/SwiftAddToWalletPlugin.swift:55]: Passing the entire channel down to the View is a bit hardcore but it's not worth much more effort yet 😅

File: lib/add_to_wallet.dart [hunk 2]
```diff
@@ -1,10 +1,39 @@
...
+    initMethodCallHandler();
+  }
+
+  void initMethodCallHandler() => _channel.setMethodCallHandler(_handleCalls);
+
+
+  Future<dynamic> _handleCalls(MethodCall call) async {
+    var handler = _handlers[call.arguments['key']];
+    return handler != null ? await handler(call) : null;
+  }
```
Comment (malkovro) [inline lib/add_to_wallet.dart:28]: Assuming the only message sent is that the button got pressed. If we had more events communication between the native and dart layer we will need to segregate over call.method, create an enum of events, etc...
Once again, at this level it's probably not worth it.

File: lib/widgets/add_to_wallet_button.dart [hunk 3]
```diff
@@ -1,30 +1,58 @@
...
-class AddToWalletButton extends StatelessWidget {
+class AddToWalletButton extends StatefulWidget {
+  static const viewType = 'PKAddPassButton';
+
   final List<int> pkPass;
   final double width;
   final double height;
   final Widget? unsupportedPlatformChild;
+  final FutureOr<void> Function()? onPress;
+  final String _id = Random.secure().nextDouble().toString();
```
Comment (malkovro) [inline lib/widgets/add_to_wallet_button.dart]: Generating a pseudo random string to identify the widget. We are not supposed to have a screen with more than a AddToWallet button  so it should not be a problem that uniqueness is not guaranteed here.

File: example/lib/main.dart [hunk 4]
```diff
@@ -56,6 +56,9 @@ class _MyAppState extends State<MyApp> {
                   width: 150,
                   height: 30,
                   unsupportedPlatformChild: DownloadPass(pkPass: _pkPassData),
+                  onPress: () {
```
Comment (David-Mou) [inline example/lib/main.dart]: Flutter is calling `onPressed` this action in their buttons. Should we go with that `standard` way if we think to make this public?

File: lib/widgets/add_to_wallet_button.dart (hunk 3 above)
Comment (David-Mou) [reply lib/widgets/add_to_wallet_button.dart]: Agree. Just thinking in release this as public, we can think in a screen that have a list of items where each item can be added to wallet. We can just generate a UUIDv4 for that, but I think that is ok meantime.

File: lib/add_to_wallet.dart [hunk 5]
```diff
@@ -1,10 +1,39 @@
...
+
+  factory AddToWallet() {
+    return _instance;
+  }
+
+  AddToWallet._internal() {
+    initMethodCallHandler();
+  }
+
+  void initMethodCallHandler() => _channel.setMethodCallHandler(_handleCalls);
```
Comment (David-Mou) [inline lib/add_to_wallet.dart]: This should be private 
```suggestion
  void _initMethodCallHandler() => _channel.setMethodCallHandler(_handleCalls);
```

**Review by David-Mou — APPROVED** [review]

File: lib/widgets/add_to_wallet_button.dart (hunk 3 above)
Comment (malkovro) [reply lib/widgets/add_to_wallet_button.dart]: I switch to Uuid as it's not too much effort

**Pull request merged**